src/main/java/org/broadinstitute/hellbender/utils/iterators/SamReaderQueryingIterator.java

```diff
@@ -0,0 +1,160 @@
+package org.broadinstitute.hellbender.utils.iterators;
+
+import htsjdk.samtools.QueryInterval;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.util.CloseableIterator;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+
+/**
+ * An iterator that allows for traversals over a SamReader restricted to a set of intervals, unmapped reads,
+ * or both. Reads overlapping the provided set of intervals will be returned first (if requested), followed
+ * by the unmapped reads with no mapped mates (if requested).
+ *
+ * Note that we can't use something like com.google.common.collect.Iterators.concat() for this purpose
+ * because of the restriction in the SamReader API that "Only a single open iterator on a given SAMFileReader
+ * may be extant at any one time.  If you want to start a second iteration, the first one must be closed first."
+ * This custom iterator ensures that this invariant is maintained.
+ */
+public class SamReaderQueryingIterator implements CloseableIterator<SAMRecord>, Iterable<SAMRecord> {
+    protected static final Logger logger = LogManager.getLogger(SamReaderQueryingIterator.class);
+
+    private final SamReader reader;
+    private final QueryInterval[] queryIntervals;
+    private final boolean queryUnmapped;
+    private CloseableIterator<SAMRecord> currentIterator;
+    private SAMRecord nextRecord;
+    private boolean intervalQueryPerformed;
+    private boolean unmappedQueryPerformed;
+
+    /**
+     * Create a SamReaderQueryingIterator given a SamReader, query intervals, and a flag indicating whether unmapped
+     * reads with no mapped mate should be returned. The query intervals may be null or empty, indicating that no
+     * mapped reads are to be returned. Unmapped reads assigned the positions of their mapped mates will be returned
```

Comment:
We should make it clear that it's not exactly unmapped reads with no mate, it's unmapped reads with no positions.